zeebe/engine/src/main/java/io/camunda/zeebe/engine/metrics/DistributionMetrics.java

```diff
@@ -0,0 +1,152 @@
+  }
+
+  public void addPendingDistribution(final int partitionId, final long distributionKey) {
+    getPartitionMetrics(partitionId).addPendingDistribution(distributionKey);
+  }
+
+  public void removePendingDistribution(final int partitionId, final long distributionKey) {
+    getPartitionMetrics(partitionId).removePendingDistribution(distributionKey);
+  }
+
+  /**
+   * This method is called when a pending distribution is not found for a received acknowledgement.
+   * This may happen if a distribution has been retried, but in the meantime an acknowledgement was
+   * already received - thus this was duplicate acknowledgement.
+   *
+   * <p>TBD if this metrics brings additional value, compared to the retryInflightDistribution
+   * metric as it also represents any delay/issues between origin and target partition.
+   *
+   * @param partitionId the partition id of the distribution
+   * @param distributionKey the key of the distribution
+   */
+  public void noPendingDistribution(final int partitionId, final long distributionKey) {
+    getPartitionMetrics(partitionId).noPendingDistribution(distributionKey);
+  }
+
+  public void addInflightDistribution(final int partitionId, final long distributionKey) {
+    getPartitionMetrics(partitionId).addInflightDistribution(distributionKey);
+  }
+
+  public void removeInflightDistribution(final int partitionId, final long distributionKey) {
+    getPartitionMetrics(partitionId).removeInflightDistribution(distributionKey);
+  }
+
+  /**
+   * This method is called on the target partition when a distribution command has been successfully
+   * processed
+   *
+   * @param partitionId the partition id of the distribution
+   * @param distributionKey the key of the distribution
+   */
+  public void acknowledgeDistribution(final int partitionId, final long distributionKey) {}
+
+  /**
+   * This method is called when a current inflight distribution is being retried. This may happen if
+   * the acknowledgement of the target partition is not received by the origin partition in-time.
+   * See CommandRedistribution.java
+   *
+   * @param partitionId the partition id of the distribution
+   * @param distributionKey the key of the distribution
+   */
+  public void retryInflightDistribution(final int partitionId, final long distributionKey) {
+    getPartitionMetrics(partitionId).retryInflightDistribution(distributionKey);
+  }
+
+  public PartitionDistributionMetrics getPartitionMetrics(final int partitionId) {
+    return partitionMetrics.computeIfAbsent(
+        partitionId, id -> new PartitionDistributionMetrics(id, meterRegistry));
+  }
+
+  public void timedRedistribution(final Runnable redistribution) {
+    // timer
+    redistribution.run();
+  }
+
+  private class PartitionDistributionMetrics {
+
+    private final MeterRegistry meterRegistry;
+    private final int partitionId;
+
+    public PartitionDistributionMetrics(final int partitionId, final MeterRegistry meterRegistry) {
```

Comment:
:upside_down_face: The meter registry given here already has a common partition ID tag, so you can assume _all_ your metrics will have a partition tag added, so you don't need to track it.

Reply:
Not exactly, this is the partition id of the target partition 😄 the one in the MeterRegistry is the id of the source/origin partition. This way we see how many outgoing distributions per target partition we have, e.g. when partition X is piling up, but others are processing distributions well.

Sure, it's up for discussion, if we need this granularity 🤷

Reply:
Ah, right, that wasn't clear to me but it makes perfect sense :)

Reply:
naming is hard, maybe we better call it `targetPartitionId` or `receiverPartitionId` 🙃